contrib/guix/libexec/build.sh

```diff
@@ -0,0 +1,380 @@
+                                   guix_ldflags="$HOST_LDFLAGS"
+
+# Log the depends package hashes
+DEPENDS_PACKAGES="$(make -C contrib/depends --no-print-directory HOST="$HOST" print-all_packages)"
+DEPENDS_CACHE="$(make -C contrib/depends --no-print-directory ${BASE_CACHE+BASE_CACHE="$BASE_CACHE"} print-BASE_CACHE)"
+
+{
+    for package in ${DEPENDS_PACKAGES}; do
+        cat "${DEPENDS_CACHE}/${HOST}/${package}"/*.hash
+    done
+} | sort -k2 > "${LOGDIR}/depends-packages.txt"
+
+# Stop here if we're only building depends packages
+if [[ -n "$DEPENDS_ONLY" ]]; then
+    exit 0
+fi
+
+###########################
+# Source Tarball Building #
+###########################
+
+# Use COMMIT_TIMESTAMP for the source and release binary archives
+export SOURCE_DATE_EPOCH=${COMMIT_TIMESTAMP}
+export TAR_OPTIONS="--owner=0 --group=0 --numeric-owner --mtime='@${SOURCE_DATE_EPOCH}' --sort=name"
+
+GIT_ARCHIVE="${DIST_ARCHIVE_BASE}/monero-source-${VERSION}.tar.gz"
+
+# Create the source tarball if not already there
+if [ ! -e "$GIT_ARCHIVE" ]; then
+    mkdir -p "$(dirname "$GIT_ARCHIVE")"
+    git ls-files --recurse-submodules \
+    | cat \
+    | sort \
+    | tar --create --transform "s,^,monero-source-${VERSION}/," --mode='u+rw,go+r-w,a+X' --files-from=- \
+    | gzip -9n > ${GIT_ARCHIVE}
+    sha256sum "$GIT_ARCHIVE"
+fi
+
+###########################
+# Binary Tarball Building #
+###########################
+
+# CFLAGS
+HOST_CFLAGS="-O2"
+HOST_CFLAGS+=$(find /gnu/store -maxdepth 1 -mindepth 1 -type d -exec echo -n " -ffile-prefix-map={}=/usr" \;)
+case "$HOST" in
+    *linux-gnu*)  HOST_CFLAGS+=" -ffile-prefix-map=${PWD}=." ;;
+    *mingw*)  HOST_CFLAGS+=" -fno-ident" ;;
+    *darwin*) unset HOST_CFLAGS ;;
+    *android*) unset HOST_CFLAGS ;;
+esac
+
+# CXXFLAGS
+HOST_CXXFLAGS="$HOST_CFLAGS"
+
+case "$HOST" in
+    arm-linux-gnueabihf) HOST_CXXFLAGS="${HOST_CXXFLAGS} -Wno-psabi" ;;
+esac
+
+export GIT_DISCOVERY_ACROSS_FILESYSTEM=1
+export USE_DEVICE_TREZOR_MANDATORY=1
+
+# Make $HOST-specific native binaries from depends available in $PATH
+export PATH="${BASEPREFIX}/${HOST}/native/bin:${PATH}"
+mkdir -p "$DISTSRC"
+(
+    cd "$DISTSRC"
+
+    # Extract the source tarball
+    tar --strip-components=1 -xf "${GIT_ARCHIVE}"
+
+    # Setup the directory where our Monero build for HOST will be
+    # installed. This directory will also later serve as the input for our
+    # binary tarballs.
+    INSTALLPATH="${DISTSRC}/installed/${DISTNAME}"
+    mkdir -p "${INSTALLPATH}"
+
+    # Configure this DISTSRC for $HOST
+    # shellcheck disable=SC2086
+    env CFLAGS="${HOST_CFLAGS}" CXXFLAGS="${HOST_CXXFLAGS}" \
+    cmake --toolchain "${BASEPREFIX}/${HOST}/share/toolchain.cmake" -S . -B build \
+      -DCMAKE_INSTALL_PREFIX="${INSTALLPATH}" \
+      -DCMAKE_EXE_LINKER_FLAGS="${HOST_LDFLAGS}" \
+      -DCMAKE_SHARED_LINKER_FLAGS="${HOST_LDFLAGS}" \
+      -DCMAKE_SKIP_RPATH=ON \
+      -DMANUAL_SUBMODULES=1
```

Comment:
We can't actually check if submodules are checked out here because we're building in an extracted source archive.

The `guix-build` script runs `git submodule update --init --recursive --progress` before starting a build to make sure submodules are checked out.